pkg/operator/controller/status.go

```diff
@@ -88,109 +86,93 @@ func (r *reconciler) syncOperatorStatus() error {
 	failingCondition := &configv1.ClusterOperatorStatusCondition{
 		Type:   configv1.OperatorFailing,
 		Status: configv1.ConditionUnknown,
 	}
 	if ns == nil {
 		failingCondition.Status = configv1.ConditionTrue
 		failingCondition.Reason = "NoNamespace"
-		failingCondition.Message = "router namespace does not exist"
+		failingCondition.Message = "ingress controller namespace does not exist"
 	} else {
 		failingCondition.Status = configv1.ConditionFalse
 	}
-	conditions = setStatusCondition(conditions, failingCondition)
+	conditions = setOperatorStatusCondition(conditions, failingCondition)
 
 	progressingCondition := &configv1.ClusterOperatorStatusCondition{
 		Type:   configv1.OperatorProgressing,
 		Status: configv1.ConditionUnknown,
 	}
 	numIngresses := len(ingresses)
-	numDeployments := len(deployments)
-	if numIngresses == numDeployments {
+	ingressesAvailable := map[string]bool{}
+	for _, ing := range ingresses {
+		for _, c := range ing.Status.Conditions {
+			if c.Type == operatorv1.IngressControllerAvailableConditionType && c.Status == operatorv1.ConditionTrue {
+				ingressesAvailable[ing.Name] = true
+				break
+			}
+		}
+	}
+	if numIngresses == len(ingressesAvailable) {
 		progressingCondition.Status = configv1.ConditionFalse
 	} else {
 		progressingCondition.Status = configv1.ConditionTrue
 		progressingCondition.Reason = "Reconciling"
 		progressingCondition.Message = fmt.Sprintf(
-			"have %d ingresses, want %d",
-			numDeployments, numIngresses)
+			"Have %d ingress controllers, want %d",
+			len(ingressesAvailable), numIngresses)
 	}
-	conditions = setStatusCondition(conditions, progressingCondition)
+	conditions = setOperatorStatusCondition(conditions, progressingCondition)
 
 	availableCondition := &configv1.ClusterOperatorStatusCondition{
 		Type:   configv1.OperatorAvailable,
 		Status: configv1.ConditionUnknown,
 	}
-	deploymentsAvailable := map[string]bool{}
-	for _, d := range deployments {
-		deploymentsAvailable[d.Name] = d.Status.AvailableReplicas > 0
-	}
-	unavailable := []string{}
-	for _, ingress := range ingresses {
-		// TODO Use the manifest to derive the name, or use labels or
-		// owner references.
-		name := "router-" + ingress.Name
-		if available, exists := deploymentsAvailable[name]; !exists {
-			msg := fmt.Sprintf("no router for ingress %q",
-				ingress.Name)
-			unavailable = append(unavailable, msg)
-		} else if !available {
-			msg := fmt.Sprintf("ingress %q not available",
-				ingress.Name)
-			unavailable = append(unavailable, msg)
-		}
-	}
-	if len(unavailable) == 0 {
+	if numIngresses == len(ingressesAvailable) {
 		availableCondition.Status = configv1.ConditionTrue
 	} else {
 		availableCondition.Status = configv1.ConditionFalse
 		availableCondition.Reason = "IngressUnavailable"
-		availableCondition.Message = strings.Join(unavailable, "\n")
+		availableCondition.Message = fmt.Sprintf(
+			"Have %d ingress controllers, want %d",
+			len(ingressesAvailable), numIngresses)
```

Comment:
Should `Available` only be true when availableReplicas for all ingresses are equal to desired replicas?

Reply:
It would set `Available` to true iff every ingresscontroller has at least one replica available, and I would set `Progressing` to true iff for any ingresscontroller, the number of available replicas is not equal to the number of desired replicas.  My reasoning is that if any replicas are available, then the ingresscontroller is serving routes, but until available replicas equals desired replicas, it is still progressing towards the desired state.

Another thought though: Will we add a `Progressing` condition to ingresscontroller? If so, we could leave this code for the clusteroperator status as is with a TODO, and once ingresscontrollers have a `Progressing` condition, the clusteroperator could reflect that: if all ingresscontrollers are available, then the clusteroperator is available; and if any ingresscontrollers are progressing, then the clusteroperator is progressing.  What do you think?

Reply:
> It would set Available to true iff every ingresscontroller has at least one replica available, and I would set Progressing to true iff for any ingresscontroller, the number of available replicas is not equal to the number of desired replicas.

I believe this is best represented as an `IngressController` status condition. We can then expose the condition through the `ClusterOperator` `Progressing` status, allowing us to decouple `Deployment` logic from `ClusterOperator` status.

> Will we add a Progressing condition to ingresscontroller?

That's my understanding. After talking with @ironcladlou, we thought it made sense to get one status condition right at a time. I'll make a TODO as a reminder of what needs to be done when the `Progressing` `IngressController status condition is implemented.